rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+`FileWatcher` that stashes the inital size and periodically checks the file
+metadata (mtime and size) to see if it should start reading at what was the end.
+
+### General tweaks
+
+In addition to extracting these various subcomponents, there are some other
+relatively simple changes we can make to help address the issues we discussed.
+
+#### File discovery and checkpointing
+
+We currently use the outdated `glob_minimum_cooldown` config option to determine
+how often to do both of these tasks. We should switch them to their own
+independent config options and allow them to be disabled (e.g. a batch use case
+does not need to continuously look for new files).
+
+We should also move them into their own periodic tasks, outside of the main read
+loop. This should give us better performance and help avoid bad behavior in
+situations where either is expensive (e.g. discovering millions of files on
+EBS).
+
+#### Read concurrency
+
+With most of the other concerns pulled out, we can afford to adjust the main
+read loop to allow for some new capabilities. The most interesting is concurrent
+reads, but it will require a bit of experimentation before we're able to
+determine if it's worthwhile. There are a few possible approaches:
+
+1. Dispatch reads to an explicit threadpool
+1. Spawn a limited number of blocking tokio tasks
+1. Implement something with `iouring`
+
+The first two both introduce the questions of sizing and the ability of the
+underlying file system to enable concurrent access in a way that actually adds
+performance. We would need to test a wide variety of scenarios to evaluate the
+best path for either.
+
+Much more interesting, but also limited, would be to build on top of `iouring`.
+Given that it's only available in modern Linux, it could not be the only
+implementation. But since modern Linux is the majority of Vector usage, it could
+be enough to cover the most demanding use cases. This approach would let us
+avoid any questions of thread counts and defer concurrency to the kernel, which
+is much better equipped to make use of the available hardware.
+
+Given these options, I would currently propose that we wait. There are other
+changes in this RFC that should impact performance positively, and we'll have
+a better view of the potential cost/benefit of these approaches once those have
+landed.
+
+#### Compression-aware fingerprints
+
+As it stands, both `Fingerprinter` and `FileWatcher` instances read data from
+files, one to checksum that data and the other to return lines. The problem is
+that only `FileWatcher` handles compression, so our fingerprints can get
+confused if a file is rotated and compressed.
+
+To address that, we should evolve `FileWatcher` into a more general wrapper
+struct for file handles. This will allow us to encapsulate all direct file
+access to within the struct, where we can more easily ensure that concerns like
+compression are handled uniformly.
+
+#### Batch mode
```

Comment:
The issue is the definition of "finished". In most cases, you do not want the source to stop when it has read all of the available data because you expect more to be written shortly. In cases where you do, having a different mode lets you express that intent.